Refactor API group validation and error reporting

bart0sh opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This is a fix for validation error messages produced by the `APIEnablementOptions.Validate`. Error message for unknown API groups didn't contain versions, which could be confusing. Here is an example of the latest one from this [failing ci-node-e2e-crio-cgrpv2-dra-1-34 job](https://storage.googleapis.com/kubernetes-ci-logs/logs/ci-node-e2e-crio-cgrpv2-dra-1-34/1954534626470924288/artifacts/test-fedora-coreos-42-20250721-3-0-gcp-x86-64/services.log): 
`F0810 13:37:19.210421    2969 services.go:119] Failed to run e2e services: failed to validate ServerRunOptions: unknown api groups api`. With this fix the error would be more descriptive and easier to understand: `Failed to run e2e services: failed to validate ServerRunOptions: unknown api groups api/stable1`

Updated API group validation logic to improve error reporting. The Validate method now checks for unknown API groups using group versions and provides more descriptive error messages when invalid or unknown API groups are encountered.
Unit test cases are updated to match the new error messages.

#### Which issue(s) this PR is related to:

https://github.com/kubernetes/kubernetes/issues/133460

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

bart0sh commented:
/test pull-kubernetes-integration
/test pull-kubernetes-e2e-gce

bart0sh commented:
/test pull-kubernetes-integration
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-unit
/test pull-kubernetes-verify

bart0sh commented:
/test pull-kubernetes-integration
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-unit
/test pull-kubernetes-verify

bart0sh commented:
/test pull-kubernetes-integration
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-unit
/test pull-kubernetes-verify

siyuanfoundation commented on staging/src/k8s.io/apiserver/pkg/server/resourceconfig/helpers.go:
```diff
@@ -220,25 +220,25 @@ func getRuntimeConfigValue(overrides cliflag.ConfigurationMap, apiKey string, de
 // ParseGroups takes in resourceConfig and returns parsed groups.
-func ParseGroups(resourceConfig cliflag.ConfigurationMap) ([]string, error) {
-	groups := []string{}
+func ParseGroups(resourceConfig cliflag.ConfigurationMap) ([]*schema.GroupVersion, error) {
```
We generally do not want to change existing public functions. Can you create a new function of `ParseGroupVersions`?

bart0sh committed:
Refactor API group validation and error reporting

Updated API group validation logic to improve error reporting.
The Validate method now provides more descriptive error messages
when unknown API groups are encountered.
Unit test cases are updated to match the new error messages.

Co-authored-by: Jordan Liggitt <jordan@liggitt.net>

bart0sh commented on staging/src/k8s.io/apiserver/pkg/server/options/api_enablement.go:
```diff
@@ -70,20 +71,23 @@ func (s *APIEnablementOptions) Validate(registries ...GroupRegistry) []error {
 	if s.RuntimeConfig[resourceconfig.APIAll] == "false" && len(s.RuntimeConfig) == 1 {
 		// Do not allow only set api/all=false, in such case apiserver startup has no meaning.
-		return append(errors, fmt.Errorf("invalid key with only %v=false", resourceconfig.APIAll))
+		return append(errors, fmt.Errorf("invalid API group with only %v=false", resourceconfig.APIAll))
```
done

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/resourceconfig/helpers.go:
```diff
@@ -241,6 +241,28 @@ func ParseGroups(resourceConfig cliflag.ConfigurationMap) ([]string, error) {
 }
 
+// ParseGroupVersions takes in resourceConfig and returns parsed group versions.
+func ParseGroupVersions(resourceConfig cliflag.ConfigurationMap) ([]*schema.GroupVersion, error) {
```
```suggestion
func ParseGroupVersions(resourceConfig cliflag.ConfigurationMap) ([]schema.GroupVersion, error) {
```

bart0sh commented on staging/src/k8s.io/apiserver/pkg/server/resourceconfig/helpers.go:
```diff
@@ -241,6 +241,28 @@ func ParseGroups(resourceConfig cliflag.ConfigurationMap) ([]string, error) {
 	return groups, nil
 }
```
This way we'll lose list of group versions on the caller side and can't check if they're registered. To solve this we'd need to either change signature of the ParseGroups to return list of group versions or call ParseGroupVersions again in the `api_enablement.go:Validate`.

bart0sh commented:
/retest
